### ISSUE TITLE
Smoothing scale for stacked n(z)

### ISSUE THREAD

**Issue by aimalz** [LSSTDESC/PZDC1paper#13]
As @sschmidt23 brought up [here](https://github.com/LSSTDESC/PZDC1paper/issues/8#issuecomment-334610449), it may not be appropriate to use the default of Scott's Rule for choosing the smoothing bandwidth necessary to convert samples to a continuous distribution via KDE.  (This conversion is necessary for metrics on p_i(z), sum_i[p_i(z)], and anything that isn't a(n empirical) CDF.)  At the very least, the paper will have to discuss the choice of a smoothing bandwidth.  For that matter, `qp` uses a [Gaussian kernel for the KDE](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.gaussian_kde.html), which is also a choice we may have to justify.

**Comment by sschmidt23** [comment]
I discussed this with a few people offline and two candidate methods for choosing a bandwidth were mentioned as candidates to test:
1) For each code, run the KS test with a grid of bandwidth values and choose the bandwidth with the minimum KS value.  This assumes that the best KS value corresponds to a "good" bandwidth choice, and ignores the fact that the KS test could be somewhat insensitive to the bandwidth if the statistic is dominated by a strong feature in the distribution, e.g. the "bump" in the stacked N(z) that is evident this example BPZ dataset:
![bpzbump](https://user-images.githubusercontent.com/11219330/31408888-d7965e6a-adbe-11e7-8e3c-3054572c7733.jpg)

2) Set the bandwidth based on a measure of the mean width of the p(z) values, either via moments of the individual p(z)'s, or perhaps sigma_IQR or sigma_MAD of the point estimates.  The smaller width in e.g. the GPz individual p(z)'s would translate to a smaller bandwidth than for BPZ.

I have not yet tested either of these two scenarios.

In addition, @janewman-pitt-edu mentioned in the other thread that we may want to have the interval size over which we want to integrate should guide our bandwidth choice.  However, I'm guessing the rules of thumb for choosing this interval will gives us a similar choice between something like "Scott's rule" that is a good choice for many cases, but may not cover all cases that we wish to cover in our test (in essence, would the integration interval choice once again be too large when we look at GPz?).

**Comment by janewman-pitt-edu** [comment]
Integration spacing rules will I think basically depend on what threshold for accuracy you choose (as well as the higher-order derivatives of the curve).  

I think rather than minimizing K-S a better choice for choosing bandwidths will generally be cross-validation.  In this case, what you might do is repeatedly choose subsets of the data to train the KDE and test (rotating the test set through the set of subsets); determine the performance of not just the mean but the 1st, 2nd, 3rd and maybe 4th derivative for different bandwidths (evaluating at the test points); and look at how the RMS error in those quantities depends on bandwidth.

**Comment by janewman-pitt-edu** [comment]
Jake van der Plas might be a good person to talk about this with.

**Comment by aimalz** [comment]
Re: cross-validation, I think that scheme may result in selecting an overly broad kernel bandwidth because it will make the KDEs of all subsets look more similar.

I think it's actually still kosher to use the bandwidth that minimizes the metric, giving each p(z) method the best chance possible.  Since this paper is about unrealistically ideal conditions with the intention of introducing realistic hurdles later, it doesn't seem unreasonable to allow the optimal bandwidth to be known a priori for a test set (based on the perfectly representative training set).

**Comment by janewman-pitt-edu** [comment]
I'm fine with that.

**Comment by sschmidt23** [comment]
Looking at the results for the point estimates, with 9 code results in GPz was a clear outlier with a sigma 4x as good as the other codes.  I contacted Ibrahim to check on this, he followed the directions on the Confluence page, which did not point to the SCATTERERR subdirectory (my fault, I did not update the directory, only the README at NERSC, it has now been updated), and was still running GPz on the old data with magnitude errors correlated with the true magnitudes (this only really affects Gps, since Ibrahim runs with the errors as features, rather than as errors on features as every other code does).  He has now uploaded a new set of results, sigma_z=0.020 is good but not 4x as good as the other codes.  Most importantly, the summed N(z) has lost much of the small scale structure, and Scott's rule may not be a bad approximation now.  Here is the summed N(z) for 10% of Ibrahim's data compared with the default Scott's rule KDE spec-z.

I will need to look at all 10 codes' N(z) plots, but if they all look like this, then we may be able to get away with just using Scott's rule after all.

![updated_gpz_nz](https://user-images.githubusercontent.com/11219330/31871672-f42e3e34-b76a-11e7-99b6-a2f87d6364cd.png)

**Comment by sschmidt23** [comment]
Following discussion on the internal reviewer comments about the different amounts of small scale structure in the Figure 4 stacked p(z) comparison.  Several people have commented that the different amount of intrinsic structure in the stacked p(z) made them difficult to compare.  My original thinking was that smoothing all to the same scale eliminated the information on the small-scale structure in the individual codes.  But, Figure 1, with the individual p(z)'s, still shows this structure, so I think we can get away with smoothing everything here without losing the relevant discussion of p(z) substructure in the paper.

I made a new branch: [issue/13/nz_smoothing_scale](https://github.com/LSSTDESC/PZDC1paper/tree/issue/13/nz_smoothing_scale)
with a new version of Figure 4 where the sum p(z) is also smoothed by the same Scott's rule bandwidth that was used in the Gaussian KDE of the true redshift histogram.  This does make things *much* easier to compare side by side.  

Figure also included below
![nzsumplot_12codes_scottsrule](https://user-images.githubusercontent.com/11219330/44961590-90277f80-aec8-11e8-9cb5-0d4bdda19afe.jpg)

compare to the old version of the figure
![nzsumplot_12codes](https://user-images.githubusercontent.com/11219330/44961595-a2092280-aec8-11e8-9ff6-01b7066cb3fd.jpg)

**Comment by janewman-pitt-edu** [comment]
Much much better!

**Comment by johannct** [comment]
indeed! But now some discussion may be specifically warranted as to the exquisite match of the 4 outstanding codes, especially TrainZ and CMNN, which are near-perfect.

**Comment by sschmidt23** [comment]
Yes, the discussion will need to change, I just wanted to get confirmation that others like the new figure better before making such changes.  This was partially prompted from an internal review comment asking why TrainZ did not agree better with the true distribution (it does agree extremely well, it was just displayed poorly).  We expect CMNN and TrainZ to do well in this situation almost by design, as the color space is fully representative for CMNN, and the TrainZ N(z) is just a random subset of the test N(z), so they should match exactly, up to Poisson fluctuations due to the finite number of samples.

**Comment by sschmidt23** [comment]
I added/modified the text in the results section to reflect the changes.  With the smoothing, it's easier to see that, in addition to TrainZ and CMNN looking quite good, TPZ and FlexZBoost look good other than one small deviation in the high-z tail, with a slight over/under-prediction evident on either side of z~1.4.  I'm now wondering if this is related to the "bump" at z~1.4 shown in other codes, which I am fairly sure is due to the 4000 AA break passing through the gap between the z and y bands.  I've added text to that effect, here's the added/modified text:

"As expected, \textsc{TrainZ} is in excellent agreement with true redshift distribution: as the training sample is selected from the same underlying distribution as the test set, the redshift distributions are identical, up to Poisson fluctuations due to the finite number of sample galaxies.  \textsc{CMNN} is also in excellent agreement for similar reasons: with a representative training sample of galaxies spanning the colour-space, the sum of the colour-matched neighbour redshifts should return the true redshift distribution. \textsc{FlexZBoost} and \textsc{TPZ} also show very good agreement, with only slight departures, with an over/under-prediction in the high redshift tail of $\hat{N}(z)$ evident around $z\sim1.4$.  In fact, several of the other codes show an excess at $z \sim 1.4$, particularly the template-based codes \textsc{BPZ}, \textsc{EAZY}, and \textsc{LePhare}.
This is likely due to the $4000$ angstrom break passing through the gap between the $z$ and $y$ filters, resulting in a drastic change in $z-y$ colour for galaxies in this redshift range.  The relative dearth of strong features blue-ward of the 4000 angstrom break seems to cause problems for the redshift range that corresponds to the break passing between the two reddest filters bandpasses of a survey.
The $z\sim1.4$ feature is one of the most prominent sources of larger uncertainty in individual galaxy $p(z)$, and is readily seen in the point-estimate plots shown in Fig.~\ref{fig:pz_pointestimates} and described in the Appendix."

**Comment by janewman-pitt-edu** [comment]
with true redshift distribution: -> with the true redshift distribution:

' for the redshift range that corresponds to the break passing between the two reddest filters bandpasses of a survey.' is a bit awkward and I would try to rephrase.

'The $z\sim1.4$ feature is one of the most prominent sources of larger uncertainty in individual galaxy $p(z)$, and is readily seen in the point-estimate plots shown in Fig.~\ref{fig:pz_pointestimates} and described in the Appendix."' doesn't quite get your point across, I think.  Maybe 'Individual galaxy p(z)'s tend to be broader around z~1.4 and point estimates are more uncertain in this regime, as seen in .... '

**Comment by sschmidt23** [comment]
I actually had a paragraph about this already written at the end of the section, which I am going to move to this location in the text and merge.  The revised text is below:

 \textsc{FlexZBoost} and \textsc{TPZ} also show very good agreement, with only slight departures, with an over/under-prediction in the high redshift tail of $\hat{N}(z)$ evident around $z\sim1.4$.  In fact, several of the other codes show an excess at $z \sim 1.4$, particularly the template-based codes \textsc{BPZ}, \textsc{EAZY}, and \textsc{LePhare}.  This is likely due to the $4000\ {\rm \AA}$ break passing through the gap between the $z$ and $y$ filters, resulting in a drastic change in $z-y$ colour for galaxies in this redshift range.  With a relative dearth of strong features blue-ward of the $4000\ {\rm \AA}$ break in most galaxy SEDs, the colour change in the two reddest filter bandpasses of a survey has a large influence on the redshift determination.  The $z\sim1.4$ feature is one of the most prominent sources of larger uncertainty in individual galaxy $p(z)$.
In our sample individual galaxy $p(z)$'s tend to be broader around $z\sim1.4$ and point estimates are more uncertain in this regime, as is readily seen in the point-estimate plots shown in Fig.~\ref{fig:pz_pointestimates} and described in the Appendix.  This feature is not unique to this dataset, it is a common occurrence in photo-$z$ estimation.  The fact that similar excesses appear in Figure~\ref{fig:nz} for \textsc{ANNz2} and \textsc{METAPhoR} shows that the effect is not limited to template-based codes.  However, the lack of such a feature in the other codes shows that it is possible to eliminate the degeneracies.  Further study on this issue may provide a solution for codes that suffer from this shortcoming.
